### PR TITLE
Graph datasets

### PR DESCRIPTION
**Pull request by Matyasch**
This is a merge into the `dataset-superclass` branch! I tested both graph datasets with a dummy dataset with just a couple sentences and it looked good (didn't throw errors and there were a bunch of tensors in self._data by the end)

### CODE REVIEW

File: data_prep/graph_dataset.py [hunk 1]
```diff
@@ -1,72 +1,155 @@
...
     """
-    def __init__(self, *args, **kwargs):
-        # preprocess text
-        # compute tf.idf and pmi
-        # generate edges
-        # generate features
-        # build data object
-        raise NotImplementedError
+    def __init__(self, corpus, device):
+        self._device = device
```
Comment (rahelbeloch) [inline data_prep/graph_dataset.py:20]: no super init calls

File: data_prep/graph_dataset.py [hunk 2]
```diff
@@ -1,72 +1,155 @@
...
+        self._tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
+        self._num_classes = corpus.num_classes()
+
+        train_texts, train_labels = corpus.train_data()
+        val_texts, val_labels = corpus.val_data()
+        test_texts, test_labels = corpus.test_data()
+
+        self._raw_texts = train_texts + val_texts + test_texts
```
Comment (rahelbeloch) [inline data_prep/graph_dataset.py:28]: for what again do we need the raw texts later? Or for what other reason are you do we store the actually inside this model

File: data_prep/graph_dataset.py [hunk 3]
```diff
@@ -1,72 +1,155 @@
...
+        doc_labels = train_labels + val_labels + test_labels
+        self._labels = torch.zeros(len(iton), device=self._device) - 1
+        self._labels[:len(doc_labels)] = torch.tensor(doc_labels)
+
+        self._data = Data(edge_index=edge_index, edge_attr=edge_attr, y=self._labels)
+        self._data.train_mask = train_mask
+        self._data.val_mask = val_mask
+        self._data.test_mask = test_mask
 
+    # THIS FUNCTION IS NOT REALLY NEEDED
```
Comment (rahelbeloch) [inline data_prep/graph_dataset.py:60]: Why? I thought it would be used on the `train.get_dataloaders` to get the number of classes for the linear classification layer

File: data_prep/graph_dataset.py [hunk 4]
```diff
@@ -1,72 +1,155 @@
...
-    def preprocess(self):
+    def _preprocess(self):
         """
         Preprocesses the corpus.
         """
         raise NotImplementedError
 
     @abc.abstractmethod
-    def generate_features(self):
+    def _generate_features(self):
```
Comment (rahelbeloch) [inline data_prep/graph_dataset.py:151]: If we are not calling this method from the init method in this class, it does not need to be here in the superclass I guess.

File: data_prep/graph_dataset.py (hunk 2 above)
Comment (rahelbeloch) [reply data_prep/graph_dataset.py:28]: Ok no, let me rephrase it: in `RobertaGraphDataset.__init__`, `self._raw_texts` is accessed as property... but then in the `_preprocess` you pass it as parameter... so this property is used in different ways in the same class. What is the motivation to once access it as a property but then pass it in the preprocess?

I would suggest to just also do `self._raw_texts`  in the preprocess and not pass it as parameter

File: data_prep/pure_graph_dataset.py [hunk 5]
```diff
@@ -1,24 +1,58 @@
...
     Text Dataset used by the pure graph model.
     """
-    def __init__(self, *args, **kwargs):
-        # class super init
-        # append features to data object
-        raise NotImplementedError
+    def __init__(self, corpus, device):
+        super().__init__(corpus, device)
+        print('Generating feature matrix')
+        self._data.x = self._generate_features(self._tokens, self._raw_texts)
```
Comment (rahelbeloch) [inline data_prep/pure_graph_dataset.py]: Why passing the self._raw_texts here? can just be accessed within the generate features as the `self` object is anyways present in that method

File: data_prep/roberta_graph_dataset.py [hunk 6]
```diff
@@ -1,24 +1,67 @@
...
-    def __init__(self, *args, **kwargs):
-        # class super init
-        # append features to data object
-        raise NotImplementedError
+    def __init__(self, corpus, device):
+        super().__init__(corpus, device)
+        self._data.doc_features, self._data.word_features = self._generate_features(self._tokens, self._raw_texts)
 
-    def preprocess(self):
+    def _preprocess(self, raw_texts):
```
Comment (rahelbeloch) [inline data_prep/roberta_graph_dataset.py]: As mentioned somewhere above, I do not see the reason to pass raw_texts, if we already have the self object and can just do `self._raw_texts` in here

**Review by rahelbeloch — CHANGES_REQUESTED** [review]

**Comment by rahelbeloch** [comment]
@Matyasch Had a look at it and made some comments

File: data_prep/graph_dataset.py (hunk 1 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:20]: Why is it needed?

File: data_prep/graph_dataset.py (hunk 3 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:60]: This depends on whether at the time when the classification layer needs the number of classes we have access to the `Data` object. If yes, then we can just ask `num_classes` from there and this function is redundant. Otherwise it is needed.

File: data_prep/graph_dataset.py (hunk 4 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:151]: I thought this would similar to the num_classes thing, where we enforce a unified interface

File: data_prep/graph_dataset.py (hunk 1 above)
Comment (rahelbeloch) [reply data_prep/graph_dataset.py:20]: I think it just is code convention because you never know what fields need to be initialized in the super classes. Especially if it is from some other library (e.g. checked in the torch_geom library and there are some attributes set). So I guess just to make sure that all fields are always correctly initialized how they should in the init it is a code convention that the first thing in an init is calling the super constructur

File: data_prep/graph_dataset.py (hunk 4 above)
Comment (rahelbeloch) [reply data_prep/graph_dataset.py:151]: Ah ok, I think you only really need an interface if somebody else from outside of the class would wan to use it...

File: data_prep/graph_dataset.py (hunk 2 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:28]: You are right, updated it

File: data_prep/graph_dataset.py [hunk 7]
```diff
@@ -1,72 +1,153 @@
...
-        # compute tf.idf and pmi
-        # generate edges
-        # generate features
-        # build data object
-        raise NotImplementedError
 
+    def __init__(self, corpus, device):
+        self._device = device
+        self._tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
+        self._num_classes = corpus.num_classes()
```
Comment (matprst) [inline data_prep/graph_dataset.py]: corpus is one of the ```Data``` classes right? If so, ```num_classes``` is a property so it should be used without parenthesis

File: data_prep/graph_dataset.py [hunk 8]
```diff
@@ -1,72 +1,153 @@
...
+
+        train_texts, train_labels = corpus.train_data()
```
Comment (matprst) [inline data_prep/graph_dataset.py]: Same as for ```num_classes```

File: data_prep/graph_dataset.py [hunk 9]
```diff
@@ -1,72 +1,153 @@
...
+        val_texts, val_labels = corpus.val_data()
```
Comment (matprst) [inline data_prep/graph_dataset.py]: Same as above

File: data_prep/graph_dataset.py [hunk 10]
```diff
@@ -1,72 +1,153 @@
...
+        test_texts, test_labels = corpus.test_data()
```
Comment (matprst) [inline data_prep/graph_dataset.py]: Same :D

**Review by matprst — CHANGES_REQUESTED** [review]
Small changes. But otherwise it looks good!

File: data_prep/graph_dataset.py (hunk 7 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py]: Yep, fixed!

**Review by matprst — APPROVED** [review]
Good to me!

File: data_prep/graph_dataset.py [hunk 11]
```diff
@@ -1,72 +1,153 @@
...
-        # preprocess text
-        # compute tf.idf and pmi
-        # generate edges
-        # generate features
-        # build data object
-        raise NotImplementedError
 
+    def __init__(self, corpus, device):
+        self._device = device
+        self._tokenizer = RobertaTokenizerFast.from_pretrained("roberta-base")
```
Comment (fvlieshout) [inline data_prep/graph_dataset.py:20]: We only need the tokenizer for the RobertaGNNDataset class right? Why is it initialized here?

File: data_prep/graph_dataset.py [hunk 12]
```diff
@@ -1,72 +1,153 @@
...
+
+        print('Generate edges')
+        edge_index, edge_attr = self._generate_edges(tf_idf, tf_idf_words, pmi_score, ntoi)
+
+        print('Generate masks')
+        train_mask, val_mask, test_mask = self._generate_masks(len(train_texts), len(val_texts),
+                                                               len(test_texts), len(iton))
+
+        doc_labels = train_labels + val_labels + test_labels
+        self._labels = torch.zeros(len(iton), device=self._device) - 1
```
Comment (fvlieshout) [inline data_prep/graph_dataset.py]: torch.full(len(iton), -1, device=self._device) would be nicer

File: data_prep/graph_dataset.py (hunk 4 above)
Comment (fvlieshout) [reply data_prep/graph_dataset.py:151]: If it's not called here. Should it still be a private method?

File: data_prep/graph_dataset.py (hunk 11 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:20]: No, the `PureGraphDataset` also uses it to know what are the words of the Roberta vocabulary

File: data_prep/graph_dataset.py (hunk 12 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py]: I agree, updated

File: data_prep/roberta_graph_dataset.py [hunk 13]
```diff
@@ -1,24 +1,61 @@
+import nltk
+nltk.download('punkt')
+from nltk import word_tokenize
```
Comment (fvlieshout) [inline data_prep/roberta_graph_dataset.py:3]: Why did you choose this tokenizer? I guess we could just use any tokenizer but I was wondering if this is the same tokenizer that is used in TfidfVectorizer: https://github.com/fvlieshout/ATCS_group1/blob/01dac384363bb858e8f69343bb7df45773f36079/data_prep/graph_utils.py#L113 It should right? Otherwise the tf-idf score is computed over a different vocabulary then the actual word nodes

Comment (Matyasch) [reply data_prep/roberta_graph_dataset.py:3]: The tf.idf tokenizer is set to be `lambda x: x` which is the identity map (maps `x` to `x`). So tf.idf does not use any tokenizer since it already gets the text tokenized

Comment (Matyasch) [reply data_prep/roberta_graph_dataset.py:3]: Oh and I just chose it because we used it in practical 1 and it's better than just using `split()`

File: data_prep/roberta_graph_dataset.py [hunk 14]
```diff
@@ -1,24 +1,61 @@
...
-        raise NotImplementedError
+        features_docs = []
+        features_words = []
+        doc_embedder = RobertaModel.from_pretrained('roberta-base')
+        token_embedder = GloVe(name='840B', dim=300)
+
+        print('Generating document node features')
+        for text in self._raw_texts:
+            encodings = self._tokenizer.encode(text, truncation=True, padding=False)
+            encodings = torch.tensor(encodings, dtype=torch.long).unsqueeze(0)
```
Comment (fvlieshout) [inline data_prep/roberta_graph_dataset.py]: why did you change the squeezing, unsqueezing?

File: data_prep/graph_dataset.py (hunk 4 above)
Comment (Matyasch) [reply data_prep/graph_dataset.py:151]: It is called inside the init of the subclasses so it's still not to be used from "outside". It's not part of the interface

**Review by fvlieshout — CHANGES_REQUESTED** [review]
Most of it looks good. I added a few questions for clarification.

File: data_prep/roberta_graph_dataset.py (hunk 14 above)
Comment (Matyasch) [reply data_prep/roberta_graph_dataset.py]: Until now it was like this:
```python3
encodings = tokenizer([sentence], truncation=True, padding=False)['input_ids']
encodings = torch.tensor(encodings, dtype=torch.long)
```
Now I use `tokenizer.encode` so I don't need to access `['input_ids']`, which I think is cleaner because `encode` is the function explicitly made for getting `input_ids`. But since encode takes in a single sentence and not a batch, I have to `unsqueeze` outside of the encoder and not "inside" with `[sentence]`. Also I think `unsqueeze` is nicer as it is more explicit

Comment (fvlieshout) [reply data_prep/roberta_graph_dataset.py]: The reason I used `encodings = tokenizer([sentence], truncation=True, padding=False)['input_ids']` is because I tried both (encode and tokenizer()) and it gave me different output. The tokenizer() one gave me the same output as Rahel's model, the encode did not. That's why I opted for tokenizer()

**Review by fvlieshout — APPROVED** [review]
Matyas answered all my questions ;)

**Review by rahelbeloch — APPROVED** [review]
Fine with me too
